lib/algo/DMD.C

```diff
@@ -135,36 +184,61 @@ void DMD::train(int k)
     delete f_snapshots;
 }
 
-void
-DMD::constructDMD(const Matrix* f_snapshots,
-                  int d_rank,
-                  int d_num_procs)
+std::pair<Matrix*, Matrix*>
+DMD::computeDMDSnapshotPair(const Matrix* snapshots)
 {
+    CAROM_VERIFY(snapshots->numColumns() > 1);
+
     // TODO: Making two copies of the snapshot matrix has a lot of overhead.
     //       We need to figure out a way to do submatrix multiplication and to
     //       reimplement this algorithm using one snapshot matrix.
-    Matrix* f_snapshots_minus = new Matrix(f_snapshots->numRows(),
-                                           f_snapshots->numColumns() - 1, f_snapshots->distributed());
-    Matrix* f_snapshots_plus = new Matrix(f_snapshots->numRows(),
-                                          f_snapshots->numColumns() - 1, f_snapshots->distributed());
-
-    // Break up snapshots into snapshots_minus and snapshots_plus
-    // snapshots_minus = all columns of snapshots except last
-    // snapshots_plus = all columns of snapshots except first
-    for (int i = 0; i < f_snapshots->numRows(); i++)
+    Matrix* f_snapshots_in = new Matrix(snapshots->numRows(),
+                                           snapshots->numColumns() - 1, snapshots->distributed());
+    Matrix* f_snapshots_out = new Matrix(snapshots->numRows(),
+                                          snapshots->numColumns() - 1, snapshots->distributed());
+
+    // Break up snapshots into snapshots_in and snapshots_out
+    // snapshots_in = all columns of snapshots except last
+    // snapshots_out = all columns of snapshots except first
+    for (int i = 0; i < snapshots->numRows(); i++)
     {
-        for (int j = 0; j < f_snapshots->numColumns() - 1; j++)
+        for (int j = 0; j < snapshots->numColumns() - 1; j++)
         {
-            f_snapshots_minus->item(i, j) = f_snapshots->item(i, j);
-            f_snapshots_plus->item(i, j) = f_snapshots->item(i, j + 1);
+            f_snapshots_in->item(i, j) = snapshots->item(i, j);
+            f_snapshots_out->item(i, j) = snapshots->item(i, j + 1);
         }
-        f_snapshots_plus->item(i, f_snapshots->numColumns() - 2) =
-            f_snapshots->item(i, f_snapshots->numColumns() - 1);
+        //f_snapshots_out->item(i, snapshots->numColumns() - 2) =
+        //    snapshots->item(i, snapshots->numColumns() - 1);
```

Comment:
Lien should be uncommented

Reply:
In fact, I am not sure if this is not necessary. Isn't `j = snapshots->numColumns() - 2` included in the for-loop right before?

Reply:
Yes, you're right. It is unnecessary. It can be removed.

Reply:
Safely done.